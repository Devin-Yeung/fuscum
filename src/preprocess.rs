use std::borrow::Cow;

pub mod python;

pub trait Preprocessor {
    fn preprocess<S: AsRef<str>>(src: &S) -> Cow<str>;
}
