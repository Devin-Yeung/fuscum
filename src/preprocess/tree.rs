use ast_grep_core::matcher::KindMatcher;
use ast_grep_core::source::Edit;
use ast_grep_core::{AstGrep, Language, StrDoc};

pub struct Tree<L: Language> {
    lang: L,
    ag: AstGrep<StrDoc<L>>,
}

impl<L: Language + Copy> Tree<L> {
    pub fn new<S: AsRef<str>>(src: S, lang: L) -> Self {
        Self {
            ag: lang.ast_grep(src.as_ref()),
            lang,
        }
    }

    pub fn apply_edit_helper(&mut self, edits: Vec<Edit<String>>) -> String {
        debug_assert_ne!(edits.len(), 0);
        let mut new_content = String::new();
        let old_content = self.ag.root().root().get_text();
        let mut start = 0;
        for diff in edits {
            let range = diff.position..diff.position + diff.deleted_length;
            new_content.push_str(&old_content[start..range.start]);
            let replacement = String::from_utf8(diff.inserted_text).unwrap();
            new_content.push_str(&replacement);
            start = range.end;
        }
        // add trailing statements
        new_content.push_str(&old_content[start..]);
        new_content
    }

    pub fn apply_edits(&mut self, edits: Vec<Edit<String>>) -> &mut Self {
        match edits.len() {
            0 => self,
            _ => {
                let new_content = self.apply_edit_helper(edits);
                self.ag = self.lang.ast_grep(&new_content);
                self
            }
        }
    }

    pub fn remove_comments<S: AsRef<str>>(&mut self, kind: S) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, "");
        self.apply_edits(edits)
    }

    pub fn subst_ident<S: AsRef<str>>(&mut self, kind: S, to: &str) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits)
    }

    pub fn subst_string<S: AsRef<str>>(&mut self, kind: S, to: &str) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits)
    }

    pub fn source(&self) -> &str {
        self.ag.source()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ast_grep_language::Python;

    #[test]
    fn var_subst() {
        let mut tree = Tree::new("def f(a, b, c):\n\ta = 1", Python);
        tree.subst_ident("identifier", "v");
        assert_eq!(tree.source(), "def v(v, v, v):\n\tv = 1");
    }
}
