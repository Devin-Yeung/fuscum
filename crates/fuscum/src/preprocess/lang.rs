use crate::preprocess::tree::Tree;
use crate::preprocess::Preprocessor;
use std::borrow::Cow;
use typed_builder::TypedBuilder;

macro_rules! impl_lang_preprocessor {
    (
        $name:ident,
        lang => $lang:expr,
        ident => $identifier_token:expr,
        string => $string_token:expr,
        comment => $comment_token:expr
    ) => {
        #[derive(TypedBuilder)]
        #[builder(doc, field_defaults(default, setter(into)))]
        pub struct $name {
            #[builder(setter(
                strip_option,
                doc = "Substitute variable to a certain unified representation"
            ))]
            subst_var: Option<String>,
            #[builder(setter(
                strip_option,
                doc = "Substitute string to a certain unified representation"
            ))]
            subst_string: Option<String>,
            #[builder(default = true, setter(doc = "Remove comments from the source code"))]
            remove_comments: bool,
        }

        impl Default for $name {
            fn default() -> Self {
                Self {
                    subst_var: Some("v".to_string()),
                    subst_string: Some("\"s\"".to_string()),
                    remove_comments: true,
                }
            }
        }

        impl Preprocessor for $name {
            fn preprocess(&self, src: &str) -> Cow<str> {
                let mut tree = Tree::new(src, $lang);

                if self.remove_comments {
                    tree.remove_comments($comment_token);
                }

                self.subst_var
                    .as_ref()
                    .map(|v| tree.subst_ident($identifier_token, v));

                self.subst_string
                    .as_ref()
                    .map(|v| tree.subst_string($string_token, v));

                let src = tree
                    .source()
                    .chars()
                    .filter(|c| !c.is_whitespace())
                    .collect::<String>();
                Cow::Owned(src)
            }
        }
    };
}

impl_lang_preprocessor!(
    PythonPreprocessor,
    lang => ast_grep_language::Python,
    ident => "identifier",
    string => "string",
    comment => "comment"
);

impl_lang_preprocessor!(
    CPreprocessor,
    lang => ast_grep_language::C,
    ident => "identifier",
    string => "string_literal",
    comment => "comment"
);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn python() {
        let src = include_str!("../../../../fixtures/langs/python.py");
        let pp = PythonPreprocessor::default();
        let res = pp.preprocess(&src);
        insta::assert_snapshot!(&format!("{src}\n\n\n{res}"))
    }

    #[test]
    fn c() {
        let src = include_str!("../../../../fixtures/langs/c.c");
        let pp = CPreprocessor::default();
        let res = pp.preprocess(&src);
        insta::assert_snapshot!(&format!("{src}\n\n\n{res}"))
    }
}
