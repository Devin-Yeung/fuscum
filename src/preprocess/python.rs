use crate::preprocess::Preprocessor;
use ast_grep_core::matcher::KindMatcher;
use ast_grep_core::source::Edit;
use ast_grep_core::{AstGrep, Language, Pattern, StrDoc};
use ast_grep_language::Python;
use std::borrow::Cow;

pub struct PythonPreprocessor;
pub struct PyTree {
    ag: AstGrep<StrDoc<Python>>,
}

impl PyTree {
    pub fn new<S: AsRef<str>>(src: S) -> Self {
        Self {
            ag: Python.ast_grep(src.as_ref()),
        }
    }

    pub fn apply_edits(&mut self, edits: Vec<Edit<String>>) -> &mut Self {
        edits
            .into_iter()
            .rev()
            .fold(&mut self.ag, |root, edit| root.edit(edit).unwrap());
        self
    }

    pub fn remove_comments(&mut self) -> &mut Self {
        let pat = KindMatcher::new("comment", Python);
        let edits = self.ag.root().replace_all(&pat, "");
        self.apply_edits(edits)
    }

    pub fn subst_ident(&mut self, to: &str) -> &mut Self {
        let pat = KindMatcher::new("identifier", Python);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits) // TODO: bottleneck, it's too slow
    }

    pub fn source(&self) -> &str {
        self.ag.source()
    }
}

impl Preprocessor for PythonPreprocessor {
    fn preprocess<S: AsRef<str>>(&self, src: &S) -> Cow<str> {
        let mut tree = PyTree::new(src);
        tree.subst_ident("v").remove_comments();
        // remove all the whitespace in the source code
        let src = tree
            .source()
            .chars()
            .filter(|c| !c.is_whitespace())
            .collect::<String>();
        Cow::Owned(src)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn var_subst() {
        let mut tree = PyTree::new("def f(a, b, c):\n\ta = 1");
        tree.subst_ident("v");
        assert_eq!(tree.source(), "def v(v, v, v):\n\tv = 1");
    }

    #[test]
    fn preprocess() {
        let src = "def f(a, b, c):\n\ta = 1";
        let pp = PythonPreprocessor.preprocess(&src);
        assert_eq!(pp, "defv(v,v,v):v=1");
    }
}
