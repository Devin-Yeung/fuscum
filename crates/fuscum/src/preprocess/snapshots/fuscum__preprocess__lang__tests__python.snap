---
source: src/preprocess/lang.rs
expression: "&format!(\"{src}\\n\\n\\n{res}\")"
snapshot_kind: text
---
def foo(a, b, c): # this is comment
    print("Hello World")


defv(v,v,v):v("s")
